pkg/importer/scheduler.go

```diff
@@ -0,0 +1,425 @@
+						}
+						for _, h := range headers {
+							s := strings.Split(h, ":")
+							if len(s) < 2 {
+								continue
+							}
+							jobConf.Headers[s[0]] = s[1]
+						}
+						return j, err
+					})
+				return err
+			}, 0,
+		); err != nil {
+			slog.Error("database error", "err", err)
+			return
+		}
+
+		t := time.Now().UTC()
+
+		go func() {
+			logFileLocation := s.cfg.Importer.LogPath + "-" + jobConf.Name + "-" + t.Format(time.RFC3339) + ".log"
+			s.wg.Add(1)
+			if err := s.db.Run(
+				s.ctx,
+				func(rctx context.Context, conn *pgxpool.Conn) error {
+					const fetchSQL = `UPDATE tasks SET log_file = $2 WHERE id = $1 RETURNING id`
+					return conn.QueryRow(rctx, fetchSQL, task.ID, logFileLocation).Scan(&task.ID)
+				}, 0,
+			); err != nil {
+				slog.Error("database error", "err", err)
+			}
+			s.wg.Done()
+			var status models.Status
+			if err := s.downloader.run(downloadCtx, DownloadJob{
+				Config:         jobConf,
+				ForwardQueue:   0,
+				Presets:        s.cfg.Importer.RemoteValidatorPresets,
+				ValidationMode: s.cfg.Importer.ValidationMode,
+				Db:             s.db,
+				LogFile:        logFileLocation,
+				LogLevel:       s.cfg.Importer.LogLevel,
+			}); err != nil {
+				slog.Error("download error", "err", err)
+				if errors.Is(err, context.Canceled) {
+					status = models.ABORTED
+				} else {
+					status = models.FAILED
+				}
+			} else {
+				status = models.COMPLETED
+			}
+
+			s.wg.Add(1)
+			_, err := s.setTaskState(task.ID, status)
+			if err != nil {
+				slog.Error("setTaskState error", "err", err)
+			}
+			s.wg.Done()
+		}()
+	}
+}
+
+func (s *Scheduler) setTaskState(taskID int64, status models.Status) (*int64, error) {
+	expr := query.FieldEqInt("id", taskID)
+	builder := query.SQLBuilder{}
+	builder.CreateWhere(expr)
+
+	var updateTaskID int64
+
+	updateSQL := `UPDATE tasks SET status = $1  WHERE ` +
+		builder.WhereClause +
+		`RETURNING id`
+
+	if err := s.db.Run(
+		s.ctx,
+		func(rctx context.Context, conn *pgxpool.Conn) error {
+			return conn.QueryRow(rctx, updateSQL, status).Scan(&updateTaskID)
+		}, 0,
+	); err != nil {
+		slog.Error("database error", "err", err)
+		return nil, err
+	}
+
+	return &updateTaskID, nil
+}
+
+// Start starts the scheduler.
+func (s *Scheduler) Start() {
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-s.notify:
+				func() {
+					s.runTasks()
+					s.runCron()
+				}()
+			}
+		}
+	}()
+}
+
+// Close closes the scheduler
+func (s *Scheduler) Close() {
+	s.wg.Wait()
+	close(s.notify)
+	s.cronLock.Lock()
+	s.cron.Stop()
+	s.cronLock.Unlock()
+
+	// Stop all running tasks
+	s.runningTaskLock.Lock()
+	tasks := make([]int64, len(s.runningTasks))
+	i := 0
+	for t := range s.runningTasks {
+		tasks[i] = t
+		i++
+	}
+	s.runningTaskLock.Unlock()
+	for _, t := range tasks {
+		if err := s.AbortTask(t); err != nil {
+			slog.Error("could not abort task", "err", err)
+		}
+	}
+}
```

Comment:
Replace all locks with waiting for all goroutines to finish. There is a race condition while closing the database connection while the download handler still uses it.